### PR TITLE
JIT: break loop canonicalization into two stages

### LINKED ISSUE #70802: [libraries-pgo]: Assertion failed 'fgDominate(head, entry)' in 'System.Text.Encoding:GetByteCount(System.String,int,int):int:this' during 'Hoist loop code' (IL size 109; hash 0x10278d31; Tier1)
Forking from #70569, see that issue for details.

### PR DESCRIPTION
**Pull request by AndyAyersMS**
For a given loop, we need to separate out the true backedge, any
non-loop backedges, and any inner loop backedges so that they all
target distinct blocks.

Otherwise we may violate assumptions that the loop entry dominates
all blocks in the loop and that all backedges that reach top come
from within the loop.

This seems simplest to do with two rounds of canonicalization, one
that moves the non-loop edges, and another that moves the true backedge.

Fixes #70802.

[Diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1829320&view=ms.vss-build-web.run-extensions-tab)

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib 

No diffs from this. I have an SPMI repro case from libraries-pgo which this fixes. Will work on turning that into a test case.

**Comment by AndyAyersMS** [comment]
On the repro case, old vs new flow graphs after find loops:
<table>
<tr>
<td>


</td>
<td>

</td>
</tr>
</table>

Note how in the new case, the inner loop has just one backedge (its "true" backedge").

File: src/coreclr/jit/optimizer.cpp [hunk 1]
```diff
@@ -2904,30 +2925,119 @@ bool Compiler::optCanonicalizeLoopNest(unsigned char loopInd)
...
+//
+// Back edges incident on loop top fall into one three groups:
+// 
+// (1) Outer non-loop backedges (preds dominated by enter where pred is not in loop)
+// (2) The canonical backedge (pred == bottom)
+// (3) Nested loop backedges or nested non-loop backedges 
+//     (preds dominated by enter, where pred is in loop, pred != bottom)
+//
+// We assume dominance has already been established by loop recognition (that is,
+// anything classfied as a loop will have all backedges dominated by loop entry,
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:2947]:
```suggestion
// anything classified as a loop will have all backedges dominated by loop entry,
```

File: src/coreclr/jit/optimizer.cpp [hunk 2]
```diff
@@ -2904,30 +2925,119 @@ bool Compiler::optCanonicalizeLoopNest(unsigned char loopInd)
...
 //   loopInd - index of the loop to consider
 //
 // Returns:
 //   true if flow changes were made
 //
+// Notes:
+//
+// Back edges incident on loop top fall into one three groups:
+// 
+// (1) Outer non-loop backedges (preds dominated by enter where pred is not in loop)
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:2941]:
```suggestion
// (1) Outer non-loop backedges (preds dominated by entry where pred is not in loop)
```

File: src/coreclr/jit/optimizer.cpp [hunk 3]
```diff
@@ -2904,30 +2925,119 @@ bool Compiler::optCanonicalizeLoopNest(unsigned char loopInd)
...
+// (2) The canonical backedge (pred == bottom)
+// (3) Nested loop backedges or nested non-loop backedges 
+//     (preds dominated by enter, where pred is in loop, pred != bottom)
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:2944]:
```suggestion
//     (preds dominated by entry, where pred is in loop, pred != bottom)
```

File: src/coreclr/jit/optimizer.cpp [hunk 4]
```diff
@@ -2904,30 +2925,119 @@ bool Compiler::optCanonicalizeLoopNest(unsigned char loopInd)
...
-        return false;
+        const bool predIsInLoop = (t->bbNum <= topPredBlock->bbNum) && (topPredBlock->bbNum <= b->bbNum);
+        if (predIsInLoop || (topPredBlock == h))
+        {
+            // no action needed
+        }
+        else
+        {
+            JITDUMP("in optCanonicalizeLoop: " FMT_LP " top " FMT_BB " (entry " FMT_BB " bottom " FMT_BB ") %shas a non-loop backedge from " FMT_BB 
+                "\n",
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:2982]:
```suggestion
                "%s\n",
```
missing replacement for "need to canonicalize non-loop backedges" output

File: src/coreclr/jit/optimizer.cpp [hunk 5]
```diff
@@ -3028,122 +3139,168 @@ bool Compiler::optCanonicalizeLoop(unsigned char loopInd)
...
+    // Make sutable updates.
     //
-    BasicBlock* const origE = optLoopTable[loopInd].lpEntry;
-    if (optLoopTable[loopInd].lpTop == origE)
+    if (option == LoopCanonicalizationOption::Current)
     {
-        optLoopTable[loopInd].lpEntry = newT;
-    }
-    optLoopTable[loopInd].lpTop = newT;
+        JITDUMP("in optCanonicalizeLoop (current): " FMT_BB " is now the top of loop %d.\n", newT->bbNum,
```
Comment (BruceForstall) [inline src/coreclr/jit/optimizer.cpp:3256]:
```suggestion
        JITDUMP("in optCanonicalizeLoop (current): " FMT_BB " is now the top of loop " FMT_LP ".\n", newT->bbNum,
```

**Review by BruceForstall — APPROVED** [review]
Looks good! A few minor comments.

File: src/coreclr/jit/optimizer.cpp [hunk 6]
```diff
@@ -2904,30 +2925,119 @@ bool Compiler::optCanonicalizeLoopNest(unsigned char loopInd)
...
+        }
+        else
+        {
+            JITDUMP("in optCanonicalizeLoop: " FMT_LP " top " FMT_BB " (entry " FMT_BB " bottom " FMT_BB ") %shas a non-loop backedge from " FMT_BB 
+                "\n",
+                loopInd, t->bbNum, e->bbNum, b->bbNum, 
+                doOuterCanon ? "also " : "",
+                topPredBlock->bbNum, 
+                doOuterCanon ? "" : ": need to canonicalize non-loop backedges");
+            doOuterCanon = true;
```
[the remark this answers is not in the document]
Comment (AndyAyersMS) [reply src/coreclr/jit/optimizer.cpp:2982]: I'm kind of ambivalent about having control flow differ for DEBUG... will probably leave this as is.

**Comment by AndyAyersMS** [comment]
Native AOT timeout
```
[EXECUTION TIMED OUT]
Exit Code:-3Executor timed out after 2700 seconds and was killed
['System.Runtime.Tests' END OF WORK ITEM LOG: Command timed out, and was killed]
```

**Comment by AndyAyersMS** [comment]
Possibly related #70549, but that was supposedly fixed.

**Pull request merged**